xapian-core/weight/weight.cc

```diff
@@ -134,4 +144,27 @@ Weight::unserialise(const string &) const
     throw Xapian::UnimplementedError("unserialise() not supported for this Xapian::Weight subclass");
 }
 
+const Weight *
+Weight::create(const string & s, const Registry & reg)
+{
+    const char *p = s.c_str();
+    std::string scheme;
+
+    while (*p != ' ') {
+	if (*p == '\0') break;
+	scheme += *p;
+	p++;
+    }
+
+    if (*p == ' ')
+	return reg.get_weighting_scheme(scheme)->create_from_parameters(p++);
+    return reg.get_weighting_scheme(scheme)->create_from_parameters(p);
```

Comment:
This (a) duplicates the code needlessly and (b) doesn't actually work - `reg.get_weighting_scheme(scheme)->create_from_parameters(p++);` calls the method with the current value of `p` and **then** increments `p`.

You can just increment `p` if it points to a space, then have a single return using `p`.